### PR TITLE
raftstore: prevent stale read while merging

### PR DESCRIPTION
**Pull request by overvenus**
Local read and read index may be unsafe if commit merge is extremly slow on sibling peers.

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -175,8 +175,9 @@ impl<'a, T> ReadyContext<'a, T> {
 bitflags! {
     // TODO: maybe declare it as protobuf struct is better.
     pub struct ProposalContext: u8 {
-        const SYNC_LOG = 0b00000001;
-        const SPLIT    = 0b00000010;
+        const SYNC_LOG       = 0b00000001;
+        const SPLIT          = 0b00000010;
+        const ROLLBACK_MERGE = 0b00000100;
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: `MERGE` should also be recorded.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -943,14 +954,20 @@ impl Peer {
...
                     let ctx = ProposalContext::from_bytes(&entry.context);
-                    if ctx.contains(ProposalContext::SPLIT) {
+                    if split_to_be_updated && ctx.contains(ProposalContext::SPLIT) {
                         self.last_committed_split_idx = entry.index;
                         split_to_be_updated = false;
                     }
+                    if rollback_merge_to_be_update && ctx.contains(ProposalContext::ROLLBACK_MERGE)
+                    {
+                        self.pending_rollback_merge = true;
+                        rollback_merge_to_be_update = true;
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why?

**Comment by overvenus** [comment]
PTAL, thanks!

/run-all-tests

**Comment by overvenus** [comment]
A friendly ping.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1606,7 +1673,12 @@ impl Peer {
         }
 
         if req.get_admin_request().has_prepare_merge() {
-            self.pre_propose_merge(req)?
+            self.pre_propose_prepare_merge(req)?;
+            ctx.insert(ProposalContext::PREPARE_MERGE);
```
Comment (hicqu) [inline src/raftstore/store/peer.rs]: I think you should add `ProposalContext::MERGE` when `req.get_admin_request().has_merge()`.

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -979,14 +991,24 @@ impl Peer {
...
-                    if ctx.contains(ProposalContext::SPLIT) {
+                    if split_to_be_updated && ctx.contains(ProposalContext::SPLIT) {
                         self.last_committed_split_idx = entry.index;
                         split_to_be_updated = false;
                     }
+                    if merge_to_be_update {
+                        if ctx.contains(ProposalContext::ROLLBACK_MERGE) {
+                            // We committed rollback merge, now we can unset the pending_merge.
+                            self.pending_merge = false;
+                            merge_to_be_update = false;
```
Comment (hicqu) [inline src/raftstore/store/peer.rs]: It's not needed?

**Comment by hicqu** [comment]
rest LGTM.

**Comment by overvenus** [comment]
A friendly ping.

File: tests/failpoints_cases/test_stale_read.rs [hunk 5]
```diff
@@ -97,30 +99,63 @@ fn stale_read_during_splitting(right_derive: bool) {
...
         &region1,
         &leader1,
         &region2,
         &leader2,
+        apply_split,
+    );
+}
+
+fn must_not_stale_read(
+    cluster: &mut Cluster<NodeCluster>,
```
Comment (Hoverbear) [inline tests/failpoints_cases/test_stale_read.rs]: This test doesn't apply to `ServerCluster`? Can it be made generic?

File: tests/failpoints_cases/test_stale_read.rs [hunk 6]
```diff
@@ -97,30 +99,63 @@ fn stale_read_during_splitting(right_derive: bool) {
...
+    stale_key: &[u8],
+    old_region: &Region,
+    old_leader: &Peer,
+    new_region: &Region,
+    new_leader: &Peer,
+    fp: &str,
+) {
+    // A new value for key2.
```
Comment (Hoverbear) [inline tests/failpoints_cases/test_stale_read.rs]: For `stale_key`, not `key2` now.

File: tests/failpoints_cases/test_stale_read.rs (hunk 5 above)
Comment (overvenus) [reply tests/failpoints_cases/test_stale_read.rs]: I think `NodeCluster` is enough, the difference between them is the transport layer, `NodeCluster` uses channel, ServerCluster uses gRPC.

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -797,16 +804,20 @@ impl Peer {
...
         // if applied_index_term isn't equal to current term.
-        //
-        // There may be stale read if the old leader splits really slow,
-        // the new region may already elecetd a new leader while the old leader still think it
-        // owns the splitted range.
-
-        // TODO: It may cause read index waits a long time.
         self.get_store().applied_index_term == self.term()
+            // There may be stale read if the old leader splits really slow,
+            // the new region may already elecetd a new leader while
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: There is a typo here. :(

File: src/raftstore/store/peer.rs [hunk 8]
```diff
@@ -797,16 +804,20 @@ impl Peer {
 
     #[inline]
     fn ready_to_handle_read(&self) -> bool {
+        // TODO: It may cause read index waits a long time.
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: `It may cause read index to wait a long time.`

File: src/raftstore/store/peer.rs [hunk 9]
```diff
@@ -962,6 +973,7 @@ impl Peer {
             // leader needs to update lease and last commited split index.
             let mut lease_to_be_updated = self.is_leader();
             let mut split_to_be_updated = self.is_leader();
+            let mut merge_to_be_update = self.is_leader();
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: Seems `lease_to_be_updated`, `split_to_be_updated` and `merge_to_be_updated` are always the same, can we combine them together?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: Actually they are different, `lease_to_be_updated` depends on `Peer.ProposalQueue`, `split_to_be_updated` depends on split commands and `merge_to_be_updated` depends on merge commands.

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -979,14 +991,32 @@ impl Peer {
...
+                        if ctx.contains(ProposalContext::ROLLBACK_MERGE) {
+                            // We committed rollback merge, in order to enable read index,
+                            // we need to unset the pending_merge.
+                            //
+                            // Also local read is safe now, we renew its leader lease
+                            // as soon as possible, in the current implementation `apply_reads`
+                            // may renew the lease if there are any consistent reads.
+                            self.pending_merge = false;
+                            merge_to_be_update = false;
+                        } else if ctx.contains(ProposalContext::PREPARE_MERGE) {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: Should we care about COMMIT_MERGE here?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: No, we dont need to, COMMIT_MERGE is committed a target region, so it wont appear here.

**Comment by overvenus** [comment]
A friendly ping.

File: src/raftstore/store/peer.rs (hunk 10 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: How do you know it's definitely going to be applied?

**Comment by overvenus** [comment]
PTAL


On Jun 25, 2018, at 12:05, Jay <notifications@github.com<mailto:notifications@github.com>> wrote:


@BusyJay commented on this pull request.

________________________________

In src/raftstore/store/peer.rs<https://github.com/pingcap/tikv/pull/3116#discussion_r197675300>:

>                          self.last_committed_split_idx = entry.index;
                         split_to_be_updated = false;
                     }
+                    if merge_to_be_update {
+                        if ctx.contains(ProposalContext::ROLLBACK_MERGE) {
+                            // We committed rollback merge, in order to enable read index,
+                            // we need to unset the pending_merge.
+                            //
+                            // Also local read is safe now, we renew its leader lease
+                            // as soon as possible, in the current implementation `apply_reads`
+                            // may renew the lease if there are any consistent reads.
+                            self.pending_merge = false;
+                            merge_to_be_update = false;
+                        } else if ctx.contains(ProposalContext::PREPARE_MERGE) {


How do you know it's definitely going to be applied?

**Comment by overvenus** [comment]
PTAL

File: src/raftstore/store/peer.rs [hunk 11]
```diff
@@ -1050,6 +1080,24 @@ impl Peer {
...
+    pub fn apply_prepare_merge(&mut self, region: metapb::Region, state: MergeState) {
+        // Set pending_merge to true, in case it reboots and a committed
+        // prepare merge is not applied yet.
+        self.pending_merge_idx = Some(state.get_commit());
+
+        self.pending_merge_state = Some(state);
+        self.set_region(region);
+    }
+
+    pub fn apply_rollback_merge(&mut self, region: Option<metapb::Region>) {
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: So when `region = None` it is intended that the region is not set?

Comment (overvenus) [reply src/raftstore/store/peer.rs]: It means no need to update `region`.

File: src/raftstore/store/peer.rs [hunk 12]
```diff
@@ -985,14 +1008,22 @@ impl Peer {
...
                         self.last_committed_split_idx = entry.index;
                         split_to_be_updated = false;
                     }
+                    if merge_to_be_update && ctx.contains(ProposalContext::PREPARE_MERGE) {
+                        // We committed prepare merge, to prevent unsafe read index,
+                        // we must set the pending_merge.
+                        debug!("commit prepare_merge at {}", entry.get_index());
+                        self.last_committed_prepare_merge_idx = entry.get_index();
+                        // To prevent unsafe local read, we suspect its leader lease.
+                        self.leader_lease.suspect(monotonic_raw_now());
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why doesn't split have to do it?

File: tests/failpoints_cases/test_stale_read.rs [hunk 13]
```diff
@@ -95,62 +97,95 @@ fn stale_read_during_splitting(right_derive: bool) {
...
+    cluster: &mut Cluster<NodeCluster>,
+    stale_key: &[u8],
+    old_region: &Region,
+    old_leader: &Peer,
+    new_region: &Region,
+    new_leader: &Peer,
+    fp: &str,
+) {
+    // A new value for stale_key.
+    let v3 = b"v3";
```
Comment (BusyJay) [inline tests/failpoints_cases/test_stale_read.rs]: Can L88~L98 be removed?

File: src/raftstore/store/peer.rs [hunk 14]
```diff
@@ -272,7 +272,7 @@ pub struct Peer {
 
     // Indicates the region is going to be merged.
     // It is set to true if the prepare merge is committed in the current term.
-    pending_merge: bool,
+    last_committed_prepare_merge_idx: u64,
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: Please update the comment.

**Comment by overvenus** [comment]
PTAL, thanks!

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**